### PR TITLE
add spring boot deploy support

### PR DESCRIPTION
**Pull request by tanhe123**
1. fun deploy spring boot project directly
2. repackage spring boot jar when code size is larger than 50M
3. fix zip not package empty folder bug

### CODE REVIEW

File: lib/frameworks/spring-boot.js [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+                jarFiles.push(relative);
+              }
+            }
+            
+            if (jarFiles.length === 0) {
+              throw new Error('could not find any jar from target folder');
+            }
+
+            if (jarFiles.length > 1) {
+              throw new Error('found more than one jar files from target folder');
```
Comment (vangie) [inline lib/frameworks/spring-boot.js]: target 加个引号

File: lib/frameworks/spring-boot.js [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            for (const file of targetContents) {
+              if (_.endsWith(file, '.jar')) {
+                const absFile = path.join(targetPath, file);
+                const relative = path.relative(codeDir, absFile);
+                jarFiles.push(relative);
+              }
+            }
+            
+            if (jarFiles.length === 0) {
+              throw new Error('could not find any jar from target folder');
```
Comment (vangie) [inline lib/frameworks/spring-boot.js]: target 加个引号

File: lib/fc.js [hunk 3]
```diff
@@ -621,28 +631,165 @@ async function processNasMappingsAndEnvs({ tpl, tplPath, runtime, codeUri, baseD
...
+
+async function detectJarfilePathFromBootstrap(bootstrapContent) {
+  const matched = bootstrapContent.match(BOOTSTRAP_SPRING_BOOT_JAR_REGEX);
+  let repackaged = false;
+  let jarfilePath;
+  
+  if (matched) {
+    jarfilePath = matched[3];
+  } else {
+    if (_.includes(bootstrapContent, 'org.springframework.boot.loader.PropertiesLauncher')) {
```
[the remark this answers is not in the document]
Comment (tanhe123) [reply lib/fc.js]: 默认的 JarLauncher 硬编码从 BOOT-INF/lib 下读，PropertiesLauncher 支持 CLASSPATH，也支持自定义一些路径

File: lib/frameworks/spring-boot.js (hunk 2 above)
Comment (tanhe123) [reply lib/frameworks/spring-boot.js]: done

File: lib/frameworks/spring-boot.js (hunk 1 above)
Comment (tanhe123) [reply lib/frameworks/spring-boot.js]: done

File: lib/fc.js [hunk 4]
```diff
@@ -621,28 +631,165 @@ async function processNasMappingsAndEnvs({ tpl, tplPath, runtime, codeUri, baseD
...
+    const data = await zip.readZipFile(jarfilePath, 'META-INF/MANIFEST.MF');
+    const content = data.toString();
+  
+    return _.includes(content, 'Spring-Boot-Version');
+  } catch (e) {
+    return false;
+  }
+}
+
+async function repackSpringBootJar(unzippedPath, absJarfilePath, targetAbsPath) {
```
[the remark this answers is not in the document]
Comment (tanhe123) [reply lib/fc.js]: done

File: lib/fc.js [hunk 5]
```diff
@@ -59,6 +61,8 @@ const defaultNasConfig = {
   MountPoints: []
 };
 
+const BOOTSTRAP_SPRING_BOOT_JAR_REGEX = new RegExp('(java .*?)-jar (.*?) ([0-9a-zA-Z./_-]+\\.jar)', 'm');
```
[the remark this answers is not in the document]
Comment (tanhe123) [reply lib/fc.js]: done

**Pull request merged**